src/h3lib/lib/geodesic_iterator.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2025 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @file geodesic_iterator.c
+ * @brief Manages geodesic polygon iterator lifecycle and traversal helpers.
+ */
+
+#include "geodesic_iterator.h"
+
+#include <stdio.h>
+
+#include "geodesic_polygon_internal.h"
+#include "h3Index.h"
+
+#define H3_CHECK(h3_call, iter_ptr)              \
+    do {                                         \
+        H3Error _err = (h3_call);                \
+        if (_err != E_SUCCESS) {                 \
+            iterErrorPolygonCompact(iter, _err); \
+            return;                              \
+        }                                        \
+    } while (0)
+
+static GeodesicPolygon *_getOrCreateGeodesicPolygon(
+    IterCellsPolygonCompact *iter) {
+    if (iter->geodesicPoly == NULL) {
+        if (iter->_polygon->geoloop.numVerts <= 0) {
+            iterErrorPolygonCompact(iter, E_DOMAIN);
+            return NULL;
+        }
+        if (iter->_polygon->numHoles > 0 && iter->_polygon->holes == NULL) {
+            iterErrorPolygonCompact(iter, E_DOMAIN);
+            return NULL;
+        }
+        for (int i = 0; i < iter->_polygon->numHoles; i++) {
+            if (iter->_polygon->holes[i].numVerts <= 0) {
+                iterErrorPolygonCompact(iter, E_DOMAIN);
+                return NULL;
+            }
+        }
+        GeodesicPolygon *poly = geodesicPolygonCreate(iter->_polygon);
+        if (!poly) {
+            iterErrorPolygonCompact(iter, E_MEMORY_ALLOC);
+            return NULL;
+        }
+        iter->geodesicPoly = poly;
```

Comment:
A lot of these cases don't seem to be tested. `numHoles` being set while `holes` is `NULL` makes sense to check although if someone mismatched the `holse` and `numHoles` there is not a lot we can do about it in general.